File: ui/robot-workbench/src-tauri/src/tests/services/agent_registry.rs
#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use crate::models::ai_agent::{
        normalize_legacy_agent_registry, AgentRegistrySettings, AgentSettings, AllAgentSettings,
        CustomAgentDefinition,
    };

    fn base_settings(model: Option<&str>) -> AgentSettings {
        AgentSettings {
            enabled: true,
            model: model.map(|value| value.to_string()),
            sandbox_mode: false,
            auto_approval: false,
            session_timeout_minutes: 30,
            output_format: "markdown".to_string(),
            debug_mode: false,
            max_tokens: None,
            temperature: None,
            transport: None,
        }
    }

    #[test]
    fn test_normalize_legacy_agent_registry_preserves_builtins_and_concurrency() {
        let mut enablement = HashMap::new();
        enablement.insert("claude".to_string(), true);
        enablement.insert("codex".to_string(), false);
        enablement.insert("gemini".to_string(), true);

        let legacy = AllAgentSettings {
            claude: base_settings(Some("claude-sonnet")),
            codex: base_settings(Some("gpt-5-codex")),
            gemini: base_settings(Some("gemini-2.5-pro")),
            autohand: base_settings(None),
            ollama: base_settings(Some("llama3.2")),
            custom_agents: vec![],
            max_concurrent_sessions: 12,
        };

        let registry = normalize_legacy_agent_registry(&enablement, &legacy);

        assert_eq!(registry.max_concurrent_sessions, 12);
        assert_eq!(registry.agents["claude"].settings.model.as_deref(), Some("claude-sonnet"));
        assert_eq!(registry.agents["codex"].enabled, false);
        assert_eq!(registry.agents["ollama"].settings.model.as_deref(), Some("llama3.2"));
        assert!(registry.agents.contains_key("autohand"));
    }

    #[test]
    fn test_agent_registry_round_trips_custom_agents() {
        let registry = AgentRegistrySettings {
            version: 2,
            max_concurrent_sessions: 10,
            agents: HashMap::new(),
            custom_agents: vec![CustomAgentDefinition {
                id: "dev-rpc".to_string(),
                name: "Dev RPC".to_string(),
                command: "dev-rpc-cli".to_string(),
                transport: "json-rpc".to_string(),
                protocol: Some("rpc".to_string()),
                prompt_mode: "protocol".to_string(),
                supports_model: true,
                supports_output_format: false,
                supports_session_timeout: false,
                supports_max_tokens: false,
                supports_temperature: false,
                supports_sandbox_mode: false,
                supports_auto_approval: false,
                supports_debug_mode: true,
                settings: base_settings(Some("dev-model")),
            }],
        };

        let serialized = serde_json::to_string(&registry).expect("serialize registry");
        let parsed: AgentRegistrySettings =
            serde_json::from_str(&serialized).expect("deserialize registry");

        assert_eq!(parsed.custom_agents.len(), 1);
        assert_eq!(parsed.custom_agents[0].id, "dev-rpc");
        assert_eq!(parsed.custom_agents[0].transport, "json-rpc");
        assert_eq!(parsed.custom_agents[0].protocol.as_deref(), Some("rpc"));
        assert_eq!(parsed.custom_agents[0].settings.model.as_deref(), Some("dev-model"));
    }
}

File: ui/robot-workbench/src/components/settings/agent-registry.ts
import { AGENTS } from '@/components/chat/agents'

export type AgentTransportKind = 'cli-flags' | 'slash-commands' | 'json-rpc' | 'acp'
export type AgentPromptMode = 'flag' | 'slash' | 'protocol'

export interface CapabilityDescriptor {
  source: string
  hint: string
  fetchModels?: boolean
  autoFetch?: boolean
}

export interface AgentCapabilityMap {
  model?: CapabilityDescriptor | null
  output_format?: CapabilityDescriptor | null
  session_timeout_minutes?: CapabilityDescriptor | null
  max_tokens?: CapabilityDescriptor | null
  temperature?: CapabilityDescriptor | null
  sandbox_mode?: CapabilityDescriptor | null
  auto_approval?: CapabilityDescriptor | null
  debug_mode?: CapabilityDescriptor | null
}

export interface BuiltInAgentProfile {
  id: string
  label: string
  shortLabel: string
  description: string
  command: string
  transport: AgentTransportKind
  promptMode: AgentPromptMode
  protocol?: 'rpc' | 'acp' | 'hybrid'
  capabilities: AgentCapabilityMap
  specialView?: 'autohand'
}

export interface CustomAgentDefinition {
  id: string
  name: string
  command: string
  transport: AgentTransportKind
  protocol?: 'rpc' | 'acp'
  prompt_mode: AgentPromptMode
  supports_model: boolean
  supports_output_format: boolean
  supports_session_timeout: boolean
  supports_max_tokens: boolean
  supports_temperature: boolean
  supports_sandbox_mode: boolean
  supports_auto_approval: boolean
  supports_debug_mode: boolean
  settings: {
    enabled?: boolean
    model?: string | null
    output_format?: string
    session_timeout_minutes?: number
    max_tokens?: number | null
    temperature?: number | null
    sandbox_mode?: boolean
    auto_approval?: boolean
    debug_mode?: boolean
  }
}

export const BUILTIN_AGENT_PROFILES: BuiltInAgentProfile[] = [
  {
    id: 'autohand',
    label: 'Autohand Code',
    shortLabel: 'Autohand',
    description: 'Protocol-first agent with dedicated JSON-RPC and ACP configuration.',
    command: 'autohand',
    transport: 'json-rpc',
    promptMode: 'protocol',
    protocol: 'hybrid',
    capabilities: {},
    specialView: 'autohand',
  },
  {
    id: 'claude',
    label: 'Claude Code CLI',
    shortLabel: 'Claude',
    description: 'Configured through standard Claude CLI flags.',
    command: 'claude',
    transport: 'cli-flags',
    promptMode: 'flag',
    capabilities: {
      model: {
        source: '--model',
        hint: 'Commander passes the selected model through Claude CLI flags.',
        fetchModels: true,
        autoFetch: true,
      },
    },
  },
  {
    id: 'codex',
    label: 'Codex',
    shortLabel: 'Codex',
    description: 'Configured through the standard Codex CLI flags.',
    command: 'codex',
    transport: 'cli-flags',
    promptMode: 'flag',
    capabilities: {
      model: {
        source: '--model',
        hint: 'Commander forwards model selection to the Codex agent.',
        fetchModels: true,
      },
    },
  },
  {
    id: 'gemini',
    label: 'Gemini',
    shortLabel: 'Gemini',
    description: 'Configured through Gemini CLI flags with approval mode chosen at runtime.',
    command: 'gemini',
    transport: 'cli-flags',
    promptMode: 'flag',
    capabilities: {
      model: {
        source: '--model',
        hint: 'Commander forwards model selection to the Gemini CLI.',
        fetchModels: true,
      },
    },
  },
]

export const BUILTIN_AGENT_IDS = BUILTIN_AGENT_PROFILES.map((profile) => profile.id)

export function getBuiltInAgentProfile(agentId: string): BuiltInAgentProfile | undefined {
  return BUILTIN_AGENT_PROFILES.find((profile) => profile.id === agentId)
}

export function defaultCustomAgentDefinition(): CustomAgentDefinition {
  return {
    id: '',
    name: '',
    command: '',
    transport: 'json-rpc',
    protocol: 'rpc',
    prompt_mode: 'protocol',
    supports_model: true,
    supports_output_format: false,
    supports_session_timeout: false,
    supports_max_tokens: false,
    supports_temperature: false,
    supports_sandbox_mode: false,
    supports_auto_approval: false,
    supports_debug_mode: true,
    settings: {
      enabled: true,
      model: '',
      output_format: 'markdown',
      session_timeout_minutes: 30,
      max_tokens: null,
      temperature: null,
      sandbox_mode: false,
      auto_approval: false,
      debug_mode: false,
    },
  }
}

export function customAgentCapabilities(agent: CustomAgentDefinition): AgentCapabilityMap {
  const source =
    agent.transport === 'json-rpc'
      ? 'JSON-RPC'
      : agent.transport === 'acp'
        ? 'ACP'
        : agent.transport === 'slash-commands'
          ? 'Slash commands'
          : 'CLI flags'

  const capability = (enabled: boolean, label: string): CapabilityDescriptor | null =>
    enabled
      ? {
          source,
          hint: `${label} is configured through the selected ${source.toLowerCase()} transport.`,
        }
      : null

  return {
    model: capability(agent.supports_model, 'Model'),
    output_format: capability(agent.supports_output_format, 'Output format'),
    session_timeout_minutes: capability(agent.supports_session_timeout, 'Session timeout'),
    max_tokens: capability(agent.supports_max_tokens, 'Max tokens'),
    temperature: capability(agent.supports_temperature, 'Temperature'),
    sandbox_mode: capability(agent.supports_sandbox_mode, 'Sandbox mode'),
    auto_approval: capability(agent.supports_auto_approval, 'Auto approval'),
    debug_mode: capability(agent.supports_debug_mode, 'Debug mode'),
  }
}

export function defaultEnabledAgentsMap(): Record<string, boolean> {
  return Object.fromEntries(BUILTIN_AGENT_IDS.map((id) => [id, true]))
}

export function getAgentShortLabel(id: string): string {
  const builtIn = getBuiltInAgentProfile(id)
  if (builtIn) return builtIn.shortLabel
  const fromAgent = AGENTS.find((agent) => agent.id === id)
  if (fromAgent) return fromAgent.displayName
  return id
}
